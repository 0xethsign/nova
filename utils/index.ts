/** Converts Ethereum wei to gwei. */
export const gweiToWei = (gwei: string | number) => {
  return 1e9 * parseFloat(gwei.toString());
};

/** Promise that resolves after `ms` millisecond.s */
export function sleep(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/** Calls `operation`, if it throws an error it will retry `retires` times waiting `delaySeconds` between each retry. */
export function retryOperation(operation: () => any, delaySeconds: number, retries: number) {
  return new Promise((resolve, reject) => {
    return operation()
      .then(resolve)
      .catch((reason) => {
        if (retries > 0) {
          return sleep(delaySeconds)
            .then(retryOperation.bind(null, operation, delaySeconds, retries - 1))
            .then(resolve)
            .catch(reject);
        }
        return reject(reason);
      });
  });
}
