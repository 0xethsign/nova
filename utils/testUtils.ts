import { ethers } from "hardhat";
import { ContractTransaction } from "ethers";

export const wait = async (tx: Promise<ContractTransaction>) => {
  const transaction = await tx;
  const receipt = await transaction.wait();

  return { transaction, receipt };
};

export const createTestWallet = (
  rpc: string,
  key = "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80"
) => {
  return new ethers.Wallet(key, new ethers.providers.JsonRpcProvider(rpc));
};

export function createFactory<T>(name: string): T {
  const artifact = require(`../artifacts/contracts/${name}.sol/${name}.json`);
  return new ethers.ContractFactory(artifact.abi, artifact.bytecode) as any;
}

export function createOVMFactory<T>(name: string): T {
  const artifact = require(`../artifacts-ovm/contracts/${name}.sol/${name}.json`);
  return new ethers.ContractFactory(artifact.abi, artifact.bytecode) as any;
}

export async function waitForL1ToL2Tx(
  tx: Promise<ContractTransaction>,
  watcher: any
) {
  console.log("Waiting for L1 -> L2 transaction...");
  const { transaction } = await wait(tx);
  const [msgHash] = await watcher.getMessageHashesFromL1Tx(transaction.hash);
  await watcher.getL2TransactionReceipt(msgHash);
  console.log("L1 -> L2 transaction completed!");
}
