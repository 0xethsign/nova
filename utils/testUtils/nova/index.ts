import { ethers } from "hardhat";
import { BigNumber } from "ethers";

/** Computes the execHash for a request like NovaExecHashLib. */
export function computeExecHash({
  nonce,
  strategy,
  calldata,
  gasPrice,
}: {
  nonce: number;
  strategy: string;
  calldata: string;
  gasPrice: number | BigNumber;
}) {
  return ethers.utils.solidityKeccak256(
    ["uint256", "address", "bytes", "uint256"],
    [nonce, strategy, calldata, gasPrice]
  );
}

export * from "./registry";
export * from "./executionManager";
