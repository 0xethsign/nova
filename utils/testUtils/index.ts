import chai from "chai";
import chaiAsPromised from "chai-as-promised";
import { jestSnapshotPlugin } from "mocha-chai-jest-snapshot";
chai.use(jestSnapshotPlugin());
chai.use(chaiAsPromised);
chai.should();

import { ethers } from "hardhat";
import { ContractReceipt, ContractTransaction } from "ethers";

import chalk from "chalk";
import { IERC20 } from "../../typechain";
import { Interface } from "ethers/lib/utils";

/** Gets an ethers factory for a contract. T should be the typechain factory type of the contract (ie: MockERC20__factory). */
export function getFactory<T>(name: string): Promise<T> {
  return ethers.getContractFactory(name) as any;
}

/** Increases EVM time by `seconds` and mines a new block. */
export async function increaseTimeAndMine(seconds: number) {
  await ethers.provider.send("evm_increaseTime", [seconds]);
  await ethers.provider.send("evm_mine", []);
}

/** Takes a contract interface and returns an array of all function sigHashes that are not pure/view.  */
export function getAllStatefulSigHashes(contractInterface: Interface) {
  return Object.entries(contractInterface.functions)
    .filter(([, fragment]) => {
      return !fragment.constant;
    })
    .map(([sig]) => contractInterface.getSighash(sig));
}

/** Records the gas usage of a transaction, and checks against the most recent saved Jest snapshot.
 * If not in CI mode it won't stop tests (just show a console log).
 * To update the Jest snapshot run `npm run gas-changed`
 */
export async function snapshotGasCost(x: Promise<ContractTransaction>) {
  let receipt: ContractReceipt = await (await x).wait();

  try {
    receipt.gasUsed.toNumber().should.toMatchSnapshot();
  } catch (e) {
    console.log(
      chalk.red(
        "(CHANGE) " +
          e.message
            .replace("expected", "used")
            .replace("to equal", "gas, but the snapshot expected it to use") +
          " gas"
      )
    );

    if (process.env.CI) {
      return Promise.reject("reverted: Gas consumption changed from expected.");
    }
  }

  return x;
}

/** Checkpoints `user`'s ether `token` balance upon calling.
 * Returns two functions (calcIncrease and calcDecrease,
 * calling calcIncrease will return the  `user`'s new `token`
 * balance minus the starting balance. Calling calcDecrease
 * subtracts the final balance from the balance.
 * */
export async function checkpointBalance(token: IERC20, user: string) {
  const startingBalance = await token.balanceOf(user);

  async function calcIncrease() {
    const finalBalance = await token.balanceOf(user);

    return finalBalance.sub(startingBalance);
  }

  async function calcDecrease() {
    const finalBalance = await token.balanceOf(user);

    return startingBalance.sub(finalBalance);
  }

  return [calcIncrease, calcDecrease];
}

export * from "./nova";
