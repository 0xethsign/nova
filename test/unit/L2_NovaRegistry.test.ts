import { getFactory, snapshotGasCost } from "../../utils/testUtils";

import { ethers } from "hardhat";
import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";

import {
  L2NovaRegistry__factory,
  MockCrossDomainMessenger,
  MockCrossDomainMessenger__factory,
  MockERC20,
  SimpleDSGuard,
  SimpleDSGuard__factory,
  MockERC20__factory,
  L2NovaRegistry,
} from "../../typechain";

describe("L2_NovaRegistry", function () {
  let signers: SignerWithAddress[];
  before(async () => {
    signers = await ethers.getSigners();
  });

  let L2_NovaRegistry: L2NovaRegistry;
  let SimpleDSGuard: SimpleDSGuard;

  /// Mocks
  let MockETH: MockERC20;
  let MockCrossDomainMessenger: MockCrossDomainMessenger;

  const fakeStrategyAddress = "0x4200000000000000000000000000000000000069";
  const fakeExecutionManagerAddress =
    "0xDeADBEEF1337caFEBAbE1337CacAfACe1337C0dE";

  describe("constructor/setup", function () {
    it("should properly deploy mocks", async function () {
      MockETH = await (
        await getFactory<MockERC20__factory>("MockERC20")
      ).deploy();

      MockCrossDomainMessenger = await (
        await getFactory<MockCrossDomainMessenger__factory>(
          "MockCrossDomainMessenger"
        )
      ).deploy();
    });

    it("should properly deploy the registry", async function () {
      L2_NovaRegistry = await (
        await getFactory<L2NovaRegistry__factory>("L2_NovaRegistry")
      ).deploy(MockETH.address, MockCrossDomainMessenger.address);
    });

    it("should allow connecting to an execution manager", async function () {
      await L2_NovaRegistry.connectExecutionManager(
        fakeExecutionManagerAddress
      );

      await L2_NovaRegistry.L1_NovaExecutionManagerAddress().should.eventually.equal(
        fakeExecutionManagerAddress
      );
    });

    it("should properly use constructor arguments", async function () {
      // Make sure the constructor params were properly entered.
      await L2_NovaRegistry.messenger().should.eventually.equal(
        MockCrossDomainMessenger.address
      );

      await L2_NovaRegistry.ETH().should.eventually.equal(MockETH.address);
    });

    it("should contain constants that match expected values", async function () {
      // Make sure that the MIN_UNLOCK_DELAY_SECONDS is as expected.
      await L2_NovaRegistry.MIN_UNLOCK_DELAY_SECONDS().should.eventually.equal(
        300
      );
    });

    describe("simpleDSGuard", function () {
      it("should properly deploy a SimpleDSGuard", async function () {
        SimpleDSGuard = await (
          await getFactory<SimpleDSGuard__factory>("SimpleDSGuard")
        ).deploy();
      });

      it("should properly init the owner", async function () {
        const [deployer] = signers;

        await SimpleDSGuard.owner().should.eventually.equal(deployer.address);
      });

      it("should properly permit authorization for specific functions", async function () {
        await SimpleDSGuard.permitAnySource(
          L2_NovaRegistry.interface.getSighash(
            "requestExec(address,bytes,uint64,uint256,uint256,(address,uint256)[])"
          )
        );

        await SimpleDSGuard.permitAnySource(
          L2_NovaRegistry.interface.getSighash(
            "requestExecWithTimeout(address,bytes,uint64,uint256,uint256,(address,uint256)[],uint256)"
          )
        );

        await SimpleDSGuard.permitAnySource(
          L2_NovaRegistry.interface.getSighash(
            "speedUpRequest(bytes32,uint256)"
          )
        );

        await SimpleDSGuard.permitAnySource(
          L2_NovaRegistry.interface.getSighash("claimInputTokens(bytes32)")
        );

        await SimpleDSGuard.permitAnySource(
          L2_NovaRegistry.interface.getSighash("unlockTokens(bytes32,uint256)")
        );

        await SimpleDSGuard.permitAnySource(
          L2_NovaRegistry.interface.getSighash("relockTokens(bytes32)")
        );

        await SimpleDSGuard.permitAnySource(
          L2_NovaRegistry.interface.getSighash("withdrawTokens(bytes32)")
        );
      });

      it("should allow setting the owner to null", async function () {
        await SimpleDSGuard.setOwner(ethers.constants.AddressZero).should.not.be
          .reverted;

        await SimpleDSGuard.owner().should.eventually.equal(
          ethers.constants.AddressZero
        );
      });
    });

    describe("dsAuth", function () {
      it("should properly init the owner", async function () {
        const [deployer] = signers;

        await L2_NovaRegistry.owner().should.eventually.equal(deployer.address);
      });

      it("should allow connecting to the SimpleDSGuard", async function () {
        await L2_NovaRegistry.authority().should.eventually.equal(
          ethers.constants.AddressZero
        );

        await L2_NovaRegistry.setAuthority(SimpleDSGuard.address).should.not.be
          .reverted;

        await L2_NovaRegistry.authority().should.eventually.equal(
          SimpleDSGuard.address
        );
      });

      it("should allow setting the owner to null", async function () {
        await L2_NovaRegistry.setOwner(ethers.constants.AddressZero).should.not
          .be.reverted;

        await L2_NovaRegistry.owner().should.eventually.equal(
          ethers.constants.AddressZero
        );
      });
    });

    describe("requestExec", function () {
      it("allows a simple request", async function () {
        const [deployer] = signers;

        const gasLimit = 420;
        const gasPrice = 69;
        const tip = 1;

        await MockETH.approve(
          L2_NovaRegistry.address,
          gasLimit * gasPrice + tip
        );

        await snapshotGasCost(
          L2_NovaRegistry.requestExec(
            fakeStrategyAddress,
            "0x20",
            gasLimit,
            gasPrice,
            tip,
            []
          )
        ).should.not.be.reverted;

        await MockETH.allowance(
          L2_NovaRegistry.address,
          deployer.address
        ).should.eventually.equal(0);
      });

      it("allows a simple request with one input token", async function () {
        const [deployer] = signers;

        const gasLimit = 100_000;
        const gasPrice = 10;
        const tip = 1337;

        const inputTokenAmount = 500;

        await MockETH.approve(
          L2_NovaRegistry.address,
          gasLimit * gasPrice + tip + inputTokenAmount
        );

        await snapshotGasCost(
          L2_NovaRegistry.requestExec(
            fakeStrategyAddress,
            "0x20",
            gasLimit,
            gasPrice,
            tip,
            [{ l2Token: MockETH.address, amount: inputTokenAmount }]
          )
        );

        await MockETH.allowance(
          L2_NovaRegistry.address,
          deployer.address
        ).should.eventually.equal(0);
      });

      it("allows a simple request with 2 input tokens", async function () {
        const [deployer] = signers;

        const gasLimit = 100_000;
        const gasPrice = 10;
        const tip = 1337;

        const inputToken1Amount = 1000;
        const inputToken2Amount = 5000;

        await MockETH.approve(
          L2_NovaRegistry.address,
          gasLimit * gasPrice + tip + inputToken1Amount + inputToken2Amount
        );

        await snapshotGasCost(
          L2_NovaRegistry.requestExec(
            fakeStrategyAddress,
            "0x20",
            gasLimit,
            gasPrice,
            tip,
            [
              { l2Token: MockETH.address, amount: inputToken1Amount },
              { l2Token: MockETH.address, amount: inputToken2Amount },
            ]
          )
        );

        await MockETH.allowance(
          L2_NovaRegistry.address,
          deployer.address
        ).should.eventually.equal(0);
      });

      it("does not allow for more than 5 input tokens", async function () {
        await L2_NovaRegistry.requestExec(
          fakeStrategyAddress,
          "0x20",
          0,
          0,
          0,
          [
            { l2Token: ethers.constants.AddressZero, amount: 0 },
            { l2Token: ethers.constants.AddressZero, amount: 0 },
            { l2Token: ethers.constants.AddressZero, amount: 0 },
            { l2Token: ethers.constants.AddressZero, amount: 0 },
            { l2Token: ethers.constants.AddressZero, amount: 0 },
            { l2Token: ethers.constants.AddressZero, amount: 0 },
          ]
        ).should.be.revertedWith("TOO_MANY_INPUTS");
      });
    });
  });
});
