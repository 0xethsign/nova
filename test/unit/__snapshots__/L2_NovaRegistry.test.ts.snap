// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`L2_NovaRegistry claimInputTokens allows claiming input tokens for a reverted request 1`] = `66178`;

exports[`L2_NovaRegistry claimInputTokens allows claiming input tokens for an executed request 1`] = `66178`;

exports[`L2_NovaRegistry execCompleted allows completing a request that overflows gas usage 1`] = `94488`;

exports[`L2_NovaRegistry execCompleted allows completing a request with input tokens 1`] = `97286`;

exports[`L2_NovaRegistry execCompleted allows completing a resubmitted request 1`] = `101669`;

exports[`L2_NovaRegistry execCompleted allows completing a reverted request with input tokens 1`] = `97268`;

exports[`L2_NovaRegistry execCompleted allows completing a simple request 1`] = `114374`;

exports[`L2_NovaRegistry execCompleted allows completing an uncled request before it dies 1`] = `95118`;

exports[`L2_NovaRegistry relockTokens allows relocking tokens 1`] = `34534`;

exports[`L2_NovaRegistry requestExec allows a simple request with 2 input tokens 1`] = `338636`;

exports[`L2_NovaRegistry requestExec allows making a simple request 1`] = `244941`;

exports[`L2_NovaRegistry requestExec allows making a simple request with one input token 1`] = `285631`;

exports[`L2_NovaRegistry requestExecWithTimeout should allow a simple request with minimum timeout 1`] = `178521`;

exports[`L2_NovaRegistry speedUpRequest allows speeding up a request scheduled to unlock after switch 1`] = `255233`;

exports[`L2_NovaRegistry speedUpRequest allows speeding up a simple request 1`] = `255246`;

exports[`L2_NovaRegistry unlockTokens allows unlocking a valid request 1`] = `57067`;

exports[`L2_NovaRegistry unlockTokens allows unlocking a valid request with input tokens 1`] = `57067`;

exports[`L2_NovaRegistry withdrawTokens allows withdrawing from a request with input tokens 1`] = `103459`;

exports[`L2_NovaRegistry withdrawTokens allows withdrawing tokens from a simple request 1`] = `84912`;
